### DISCUSSION TITLE
Webots_SUMO Driving simulator development

### DISCUSSION THREAD

**Discussion by tian96-pu** [cyberbotics/webots#4563] [Q&A]
I am trying to develop a driving simulator using webots and SUMO. I have finished the pipeline but I would like to refine the work. Can someone help me with the following issues:
1) How to change the view of the driving into the following figure and lock the view
![image](https://user-images.githubusercontent.com/55172725/170903906-ceb5a7ef-18a4-4daa-8b2d-986ae28d2eb6.png)

Currently, my view is third person view like below:
![image](https://user-images.githubusercontent.com/55172725/170904470-29501b83-750a-4c29-968b-2fd1223cf3b7.png)

However, I tried to change into the first person view, the view does not look very good like the following picture: we can still see the boundary of the car cab.
![image](https://user-images.githubusercontent.com/55172725/170904544-1c397517-efb5-4906-8558-5fd72f4438cf.png)

2) The vioce of the vehicle engine is breaking up. Did someone meet this question before as well?

**Comment by stefaniapedrazzi** [comment]
To get an inboard camera view you have to:

1) open the context menu (right-click on the car in the scene tree)
2) select the option: "Follow object" > "Mounted Shot"
3) move the viewpoint to the desired view inside the car

Then, when you will start the simulation, the viewpoint will automatically move with the car.
To get a nice view of the interior of the car, you should carefully choose the position of the viewpoint and the `Viewpoint.near` value.

About question 2. with issue, you should provide more information about the issue and more information about your system.

**Reply by tian96-pu** [reply]
Thanks for your reply and it works by using your methods.
However, I found some new issues to be addressed as shown in the following videos:
1) The mirrors of the vehicles are white all the time and nothing is refelected. 
2) The current angle in my case and also in the screenshot I attached before is hard for the drivers to observe what is happening outsides the window. Could we display the driving simulators into three screens like shown in the figure below.

For the question 2, if you could view the video, you can hear the vehicles engine sound is breaking up during the running of the webots. I am using Razer laptop with AMD RH5900 CPU, and Nvida RTX3070 Laptop, 16GB RAM.

![Picture1](https://user-images.githubusercontent.com/55172725/172954840-43d50a1b-8237-4279-95a1-a7f03844b10a.jpg)


https://user-images.githubusercontent.com/55172725/172954867-b0eb66d4-865c-49e8-b063-b72883b2ffd4.mp4

**Reply by tian96-pu** [reply]
The mirror issues are solved by turning the options into true in the vehicles model.

**Reply by omichel** [reply]
> The current angle in my case and also in the screenshot I attached before is hard for the drivers to observe what is happening outsides the window. Could we display the driving simulators into three screens like shown in the figure below.

Yes, this possible by using robot cameras and displaying these cameras as external windows: if you open a world with a camera image overlay, like [samples/devices/camera.wbt](https://github.com/cyberbotics/webots/blob/master/projects/samples/devices/worlds/camera.wbt) and double click on the camera overlay, it will open the camera image in an external window, which you can drag on a different screen and resize to full-screen if needed. When saving, Webots will remember to display this camera image on that screen when it is opened again.

**Reply by omichel** [reply]
About the sound, it actually sounds strange indeed and I suspect the sounds are contact sounds. You can silence them to hear only the motor sounds, by setting a `ContactProperties` node in the `WorldInfo.contactProperties` and putting the contact sounds to empty strings:
```js
WorldInfo {
  ...
  contactProperties ContactProperties {
    bumpSound ""
    rollSound ""
    slideSound ""
  }
}
```

**Reply by tian96-pu** [reply]
> 

Thanks for your reply. This issue is solved.

**Reply by tian96-pu** [reply]
> > The current angle in my case and also in the screenshot I attached before is hard for the drivers to observe what is happening outsides the window. Could we display the driving simulators into three screens like shown in the figure below.
> 
> Yes, this possible by using robot cameras and displaying these cameras as external windows: if you open a world with a camera image overlay, like [samples/devices/camera.wbt](https://github.com/cyberbotics/webots/blob/master/projects/samples/devices/worlds/camera.wbt) and double click on the camera overlay, it will open the camera image in an external window, which you can drag on a different screen and resize to full-screen if needed. When saving, Webots will remember to display this camera image on that screen when it is opened again.

Thanks for your reply. I understand how to click the camera overlay and make it as an external windows. But how could I add a camera near the vehicle window. I can only find the camera device if I would like to add to in the sensorslot node under the vehicle. Also, do I need to create a new controller to make the camera move with the vehicle?

**Reply by omichel** [reply]
Yes, you should put the cameras in the sensor slot and simply modify the robot controller of the vehicle to enable the camera with a chosen refresh rate.
